zerver/middleware.py

```diff
@@ -328,29 +331,58 @@ def csrf_failure(request, reason=""):
 class RateLimitMiddleware(MiddlewareMixin):
     def process_response(self, request, response):
         # type: (HttpRequest, HttpResponse) -> HttpResponse
-        if not settings.RATE_LIMITING:
+        if hasattr(request, 'user'):
+            entity = RateLimitedUser(request.user)
+            response = self._process_response(entity, request, response)
+
+        if settings.IP_RATE_LIMITING:
+            ip = get_ip(request)
+            if ip:
+                response  = self._process_response(
+                    RateLimitedIP(ip), request, response, prefix='_ip')
+
+        if settings.EMAIL_RATE_LIMITING:
+            prefix = '_email'
+            attr = '{}_ratelimit_value'.format(prefix)
+            if hasattr(request, attr):
+                email = getattr(request, attr)
+                response  = self._process_response(
+                    RateLimitedEmail(email), request, response, prefix=prefix)
+
+        return response
+
+    def _process_response(self, entity, request, response, prefix=''):
+        # type: (RateLimitedObject, HttpRequest, HttpResponse, Text) -> HttpResponse
+        if not (settings.RATE_LIMITING or settings.IP_RATE_LIMITING or
+                settings.EMAIL_RATE_LIMITING):
             return response
 
-        from zerver.lib.rate_limiter import max_api_calls, RateLimitedUser
+        header_prefix = prefix.title().replace('_', '-')
         # Add X-RateLimit-*** headers
-        if hasattr(request, '_ratelimit_applied_limits'):
-            entity = RateLimitedUser(request.user)
-            response['X-RateLimit-Limit'] = str(max_api_calls(entity))
-            if hasattr(request, '_ratelimit_secs_to_freedom'):
-                response['X-RateLimit-Reset'] = str(int(time.time() + request._ratelimit_secs_to_freedom))
-            if hasattr(request, '_ratelimit_remaining'):
-                response['X-RateLimit-Remaining'] = str(request._ratelimit_remaining)
+        attr = '{}_ratelimit_applied_limits'.format(prefix)
+        if hasattr(request, attr):
+            header = 'X{}-RateLimit-Limit'.format(header_prefix)
+            response[header] = str(max_api_calls(entity))
+            attr = '{}_ratelimit_secs_to_freedom'.format(prefix)
+            if hasattr(request, attr):
+                header = 'X{}-RateLimit-Reset'.format(header_prefix)
+                response[header] = str(int(time.time() + getattr(request, attr)))
+            attr = '{}_ratelimit_remaining'.format(prefix)
+            if hasattr(request, attr):
+                header = 'X{}-RateLimit-Remaining'.format(header_prefix)
+                response[header] = str(getattr(request, attr))
         return response
 
     def process_exception(self, request, exception):
         # type: (HttpRequest, Exception) -> Optional[HttpResponse]
         if isinstance(exception, RateLimited):
+            secs_to_freedom = exception.args[0]
```

Comment:
Well, it doesn't simplify it much when we also still pass the data on the request object. :-)

Given that we're sticking data on the request object even in the case where the user is nowhere near the limit -- when there's no exception being raised -- and using that data to return metadata in the response, I think it's best to stick to that mechanism.

For identifying which ratelimit was the cause of the request aborting, we already have the `{prefix}_ratelimit_over_limit` attribute -- that's a boolean that corresponds exactly to whether the particular ratelimit aborted the request.